src/Temporalio.Extensions.DiagnosticSource/CustomMetricMeter.cs

```diff
@@ -0,0 +1,188 @@
+
+            public void Add(T value, object tags) =>
+                underlying.Add(value, ((Tags)tags).TagList);
+        }
+
+        private sealed class CustomMetricHistogram<T> : ICustomMetricHistogram<T>
+            where T : struct
+        {
+            private readonly Histogram<T> underlying;
+
+            internal CustomMetricHistogram(Histogram<T> underlying) => this.underlying = underlying;
+
+            public void Record(T value, object tags) =>
+                underlying.Record(value, ((Tags)tags).TagList);
+        }
+
+#pragma warning disable CA1001 // We are disposing the lock on destruction since this can't be disposable
+        private sealed class CustomMetricGauge<T> : ICustomMetricGauge<T>
+#pragma warning restore CA1001
+            where T : struct
+        {
+            private readonly string name;
+            // We must dedupe based on tag set. We considered several designs including regular
+            // dictionary, blocking collection or concurrent queue w/ read time dedupe, and others
+            // and it was clearest to use a concurrent dictionary for collection but lock it while
+            // collecting metrics. Iterating a concurrent dictionary does not guarantee duplicates
+            // will be prevented, so we must lock it while iterating during collection.
+            private readonly ConcurrentDictionary<Tags, Measurement<T>> pending = new();
+            private readonly ReaderWriterLockSlim pendingLock = new();
+
+            internal CustomMetricGauge(string name) => this.name = name;
+
+            ~CustomMetricGauge() => pendingLock.Dispose();
+
+            public void Set(T value, object tags)
+            {
+                // We need to support an extreme max here just to prevent unbounded memory growth.
+                // We do not want to require a logger for this, so at some extreme value we will
+                // just write to stderr and drop.
+                if (pending.Count > 50000)
+                {
+                    Console.Error.WriteLine($"Dropping gauge metric {name} since cardinality has grown over 50k");
+                    return;
+                }
+                // We are grabbing read lock because this is ok to happen concurrently even though
+                // technically we are doing more than reading
+                var tagsObj = (Tags)tags;
+                var measurement = new Measurement<T>(value, tagsObj.TagList);
+                pendingLock.EnterReadLock();
+                try
+                {
+                    pending[tagsObj] = measurement;
+                }
+                finally
+                {
+                    pendingLock.ExitReadLock();
+                }
+            }
+
+            public List<Measurement<T>> Snapshot()
+            {
+                pendingLock.EnterWriteLock();
+                try
+                {
+                    return pending.Values.ToList();
```

Comment:
Not sure what underlying's expecting here, but shouldn't `pending` be emptied once snapshot has been called?

Reply:
No, this isn't a "drain" but a "snapshot". Yes, that means that the dictionary entries never go away, so you can only have 50k tag combos for a single gauge. Basically this is asking for the gauge's current values and when it asks again, it will want the current values again.